File: app/src/main/java/com/ngengs/android/popularmovies/apps/adapters/ReviewListAdapter.java
package com.ngengs.android.popularmovies.apps.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.ngengs.android.popularmovies.apps.R;
import com.ngengs.android.popularmovies.apps.data.ReviewDetail;
import com.ngengs.android.popularmovies.apps.databinding.ItemReviewListBinding;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ngengs on 7/1/2017.
 */

@SuppressWarnings("DefaultFileTemplate")
public class ReviewListAdapter extends RecyclerView.Adapter<ReviewListAdapter.ViewHolder> {
    private static final String TAG = "ReviewListAdapter";

    private final Context context;
    private final List<ReviewDetail> data;
    private final ReviewListAdapter.ClickListener clickListener;

    public ReviewListAdapter(Context context, ClickListener clickListener) {
        this.context = context;
        this.clickListener = clickListener;
        data = new ArrayList<>();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemReviewListBinding binding = ItemReviewListBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Log.d(TAG, "onBindViewHolder: position: " + position + " size: " + getItemCount());
        ReviewDetail review = data.get(position);
        holder.binding.reviewPeople.setText(review.getAuthor());
        holder.binding.reviewText.setText(context.getString(R.string.review_text, review.getContent()));
        if (position == (getItemCount() - 1)) holder.binding.reviewSeparator.setVisibility(View.GONE);
        else holder.binding.reviewSeparator.setVisibility(View.VISIBLE);
        holder.binding.rootReview.setOnClickListener(view -> holder.itemClick());
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void add(List<ReviewDetail> data) {
        int oldSize = getItemCount();
        this.data.addAll(data);
        notifyItemRangeInserted(oldSize, data.size());
    }

    public void clear() {
        int lastSize = getItemCount();
        data.clear();
        notifyItemRangeRemoved(0, lastSize);
    }

    public ReviewDetail get(int position) {
        return data.get(position);
    }

    public List<ReviewDetail> get() {
        return data;
    }

    public interface ClickListener {
        void onClickListener(int position);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ItemReviewListBinding binding;

        ViewHolder(ItemReviewListBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void itemClick() {
            clickListener.onClickListener(getAdapterPosition());
        }
    }
}

File: app/src/main/java/com/ngengs/android/popularmovies/apps/adapters/ReviewListAdapter.kt
package com.ngengs.android.popularmovies.apps.adapters

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.ngengs.android.popularmovies.apps.R
import com.ngengs.android.popularmovies.apps.data.remote.ReviewDetail
import com.ngengs.android.popularmovies.apps.databinding.ItemReviewListBinding

/**
 * Created by rizky.kharisma on 09/11/22.
 * @ngengs
 */
class ReviewListAdapter(
    private val context: Context,
    private val clickListener: ClickListener,
    private val data: MutableList<ReviewDetail> = mutableListOf(),
) : RecyclerView.Adapter<ReviewListAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            ItemReviewListBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        Log.d(TAG, "onBindViewHolder: position: $position size: $itemCount")
        val review = data[position]
        holder.binding.reviewPeople.text = review.author
        holder.binding.reviewText.text = context.getString(R.string.review_text, review.content)
        holder.binding.reviewSeparator.visibility =
            if (position == itemCount - 1) View.GONE else View.VISIBLE
        holder.binding.rootReview.setOnClickListener { clickListener.onClickListener(position) }
    }

    override fun getItemCount(): Int = data.size

    fun add(data: List<ReviewDetail>) {
        val oldSize = itemCount
        this.data.addAll(data)
        notifyItemRangeInserted(oldSize, data.size)
    }

    fun clear() {
        val lastSize = itemCount
        data.clear()
        notifyItemRangeRemoved(0, lastSize)
    }

    fun get(position: Int): ReviewDetail? = data.getOrNull(position)
    fun get(): List<ReviewDetail> = data

    interface ClickListener {
        fun onClickListener(position: Int)
    }

    class ViewHolder(val binding: ItemReviewListBinding) : RecyclerView.ViewHolder(binding.root)

    companion object {
        private const val TAG = "ReviewListAdapter"
    }
}